### PR TITLE
Allow 65536 pages

### PR DESCRIPTION
**Pull request by k-nasa**
# Description

fix: https://github.com/wasmerio/wasmer/issues/2187

> Memory minimum and maximum sizes range 0..65536 inclusive. The verifier blocks 65537 and higher. However, wasmer only works on 65535 and lower.

I fixed this issue.

# Review

- [x] Add a short description of the change to the CHANGELOG.md file

### CODE REVIEW

**Comment by k-nasa** [comment]
I would like to add a test to detect this issue. But I don't understand structure of tests in this repository so I don't know where to add it. Can you give me a signpost?

**Comment by nlewycky** [comment]
Thank you for the patch! A `.wast` type test in `tests/wast/wasmer` should do it.

File: lib/vm/src/memory.rs [hunk 1]
```diff
@@ -259,6 +259,7 @@ impl LinearMemory {
         let mapped_pages = memory.minimum;
         let mapped_bytes = mapped_pages.bytes();
 
+        println!("{}", minimum_bytes);
```
Comment (syrusakbary) [inline lib/vm/src/memory.rs:262]: Delete debug print

Comment (k-nasa) [reply lib/vm/src/memory.rs]: 🙈  I fixed it. 755639c

**Comment by k-nasa** [comment]
Thanks @nlewycky! I was able to add test.

File: lib/types/src/units.rs [hunk 2]
```diff
@@ -91,8 +91,8 @@ impl From<usize> for Bytes {
     }
 }
 
-impl From<u32> for Bytes {
-    fn from(other: u32) -> Self {
+impl From<u64> for Bytes {
+    fn from(other: u64) -> Self {
```
Comment (Hywan) [inline lib/types/src/units.rs:95]: Please keep both implementations, otherwise it's a regression :-).

File: lib/vm/src/vmcontext.rs [hunk 3]
```diff
@@ -340,7 +340,7 @@ pub struct VMMemoryDefinition {
     pub base: *mut u8,
 
     /// The current logical size of this linear memory in bytes.
-    pub current_length: u32,
+    pub current_length: u64,
```
Comment (Hywan) [inline lib/vm/src/vmcontext.rs:343]: I wonder if we should not create a type alias to represent the memory size, something like `pub type MemorySize = u64`. Thoughts @nlewycky, @MarkMcCaskey, @syrusakbary?

Comment (nlewycky) [reply lib/vm/src/vmcontext.rs:343]: I don't think so in this case, we aren't going to be changing this type or confusing it with other values much.

**Review by nlewycky — COMMENTED** [review]
This:
https://sourcegraph.com/github.com/wasmerio/wasmer/-/blob/lib/compiler-llvm/src/translator/intrinsics.rs#L620:13
also need to be updated from `i32` to `i64` or else we'll compute the wrong value, and that may require some further changes (llvm is a strongly-typed IR and so any place it's used with another i32 will fail the verifier).

Similarly singlepass loads a 32-bit value into a register here:
https://sourcegraph.com/github.com/wasmerio/wasmer/-/blob/lib/compiler-singlepass/src/codegen_x64.rs#L1280

I'm not sure what the equivalent is in cranelift, possibly the type of the global_value in make_heap in src/func_environ.rs (not src/translator/func_environ.rs)?

**Comment by Amanieu** [comment]
Superseded by #2677

**Pull request closed without merging**